### PR TITLE
fix(certificate): hybrid dp cannot refresh certificate entity with vault reference

### PR DESCRIPTION
**Pull request by windmgc**
### Summary

This PR fixes a problem that a certificate entity with cert/keys stored in a vault-referenced type cannot be refreshed even if the vault reference secret value has been updated both in L1/L2 vault PDK.

The root cause of this problem is that in a dp node running hybrid mode, kong cache interfaces force the cache TTL to be infinite(and it also does not respect cache TTL configured in `kong.conf`), thus the certificate object cannot be updated unless a new configuration is received and cache `purge` being called.

 The PR fixes the problem by caching the vault reference fields `$ref` and call `kong.vault.update` every time in `get_certificate`, and l1 serializer will keep the value serialized inside the certificate object.
 
This PR contains a test case that tests a certificate entity with a vault-referenced key that can be refreshed promptly, according to the TTL configured on the vault reference. The test case will fail consistently without the fix.

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

FTI-5881

### CODE REVIEW

**Comment by bungle** [comment]
why wasn't this enough:
https://github.com/Kong/kong/blob/master/kong/runloop/certificate.lua#L217

I don't understand.

**Comment by windmgc** [comment]
@bungle  The value returned by this line https://github.com/Kong/kong/blob/1714e26e033ee7b0d296739149c8b19e2d96dc0e/kong/runloop/certificate.lua#L211 is something like 
```
{cert = cdata<void *>: 0x0118f09730, key = cdata<void *>: 0x0118f0be20}
```

Without the `$ref` fields and plain cert/key fields. Because it has a l1 serializer configured here https://github.com/Kong/kong/blob/1714e26e033ee7b0d296739149c8b19e2d96dc0e/kong/runloop/certificate.lua#L170

So we're actually caching these cdata objects inside the L1 cache.

Calling `kong.vault.update` on this table will have no real effect

**Comment by bungle** [comment]
> certificate object cannot be refreshed unless a new configuration is received and cache purge being called.

The certificate entity itself is not refreshed unless there is new config. In which case it is read again from LMDB (because caches are purged)? But vault.update is always called before certificate is served.

> The PR fixes the problem by overriding the TTL value returned by the certificate fetching callback function

This is where I have problem of understanding. Why would infinite value be a problem if caches are purged, and entity cannot change without caches being purged?

**Comment by bungle** [comment]
> Because it has a l1 serializer configured here

Oh... that is the key! Thanks.

**Comment by bungle** [comment]
@windmgc check last commit here for alternate proposal (basically only changing certificate.lua):
https://github.com/Kong/kong/pull/12870

**Comment by windmgc** [comment]
@bungle Please review again, thanks!

**Comment by bungle** [comment]
@kikito I added some backport labels to this.

**Review by bungle — APPROVED** [review]

File: spec/02-integration/13-vaults/05-ttl_spec.lua [hunk 1]
```diff
@@ -221,5 +222,204 @@ describe("vault ttl and rotation (#" .. strategy .. ") #" .. vault.name, functio
...
+  }
+
+  local vault_fixtures = vault:fixtures()
+  vault_fixtures.dns_mock = tls_fixtures.dns_mock
+
+  lazy_setup(function()
+    helpers.setenv("KONG_LUA_PATH_OVERRIDE", LUA_PATH)
+    helpers.setenv("KONG_VAULT_ROTATION_INTERVAL", "1")
+
+    helpers.test_conf.loaded_plugins = {
```
Comment (Water-Melon) [inline spec/02-integration/13-vaults/05-ttl_spec.lua:269]: This assignment statement will cause a [flakiness](https://konghq.atlassian.net/browse/KAG-4240). Please just remove it.

**Review by Water-Melon — CHANGES_REQUESTED** [review]

File: spec/02-integration/13-vaults/05-ttl_spec.lua (hunk 1 above)
Comment (windmgc) [reply spec/02-integration/13-vaults/05-ttl_spec.lua]: Fixed, @Water-Melon please review again

**Review by Water-Melon — APPROVED** [review]

**Pull request merged**